statsd/telemetry.go

```diff
@@ -0,0 +1,116 @@
+package statsd
+
+import (
+	"fmt"
+	"sync"
+	"time"
+)
+
+/*
+TelemetryInterval is the interval at which telemetry will be sent by the client.
+*/
+const TelemetryInterval = 10 * time.Second
+
+/*
+clientTelemetryTag is a tag identifying this specific client.
+*/
+var clientTelemetryTag = "client:go"
+
+/*
+clientVersionTelemetryTag is a tag identifying this specific client version.
+*/
+var clientVersionTelemetryTag = "client_version:3.7.2"
+
+type telemetryClient struct {
+	c         *Client
+	tags      []string
+	namespace string
+	sender    *sender
+	worker    *worker
+}
+
+func NewTelemetryClient(c *Client, transport string) *telemetryClient {
+	return &telemetryClient{
+		c:         c,
+		tags:      append(c.Tags, clientTelemetryTag, clientVersionTelemetryTag, "client_transport:"+transport),
+		namespace: c.Namespace,
+	}
+}
+
+func NewTelemetryClientWithCustomAddr(c *Client, transport string, telemetryAddr string, pool *bufferPool) (*telemetryClient, error) {
+	telemetryWriter, _, err := resolveAddr(telemetryAddr)
+	if err != nil {
+		return nil, fmt.Errorf("Could not resolve telemetry address: %v", err)
+	}
+
+	t := NewTelemetryClient(c, transport)
+
+	// creating a custom sender/worker with 1 worker in
+	// mutex mode for the telemetry that share the same
+	// bufferPool
+	t.sender = newSender(telemetryWriter, DefaultUDPBufferPoolSize, pool)
```

Comment:
I would probably add a FIXME about this default value as well. Or even the code

if (udp) {
  poolSize = default UDP
} else {
  poolSize = default UDP // FIXME due to performance pitfall, we're using UDP defaults for UDS.
}